File: src/main/java/org/opensearch/commons/ppl/TransportPPLQueryRequest.java
/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

package org.opensearch.commons.ppl;

import java.io.IOException;

import org.opensearch.action.ActionRequest;
import org.opensearch.action.ActionRequestValidationException;
import org.opensearch.core.common.io.stream.StreamInput;
import org.opensearch.core.common.io.stream.StreamOutput;

/**
 * This is a copied, reduced version of SQL Plugin's TransportPPLQueryRequest
 */
public class TransportPPLQueryRequest extends ActionRequest {

    // Ordinals must match org.opensearch.sql.protocol.response.format.JsonResponseFormatter.Style
    private enum Style {
        COMPACT,
        PRETTY
    }

    // actually used fields
    private final String pplQuery;
    private final String jsonContentString;
    private final String path;

    // fields put here to match the original TransportPPLQueryRequest
    private final String format;
    private final String explainMode;
    private final boolean sanitize;
    private final Style style;
    private final boolean profile;
    private final String queryId;

    public TransportPPLQueryRequest(String pplQuery, String jsonContentString, String path) {
        this.pplQuery = pplQuery;
        this.jsonContentString = jsonContentString;
        this.path = path;
        this.format = "";
        this.explainMode = null;
        this.sanitize = true;
        this.style = Style.COMPACT;
        this.profile = false;
        this.queryId = null;
    }

    public TransportPPLQueryRequest(StreamInput in) throws IOException {
        super(in);
        this.pplQuery = in.readOptionalString();
        this.format = in.readOptionalString();
        this.explainMode = in.readOptionalString();
        this.jsonContentString = in.readOptionalString();
        this.path = in.readOptionalString();
        this.sanitize = in.readBoolean();
        this.style = in.readEnum(Style.class);
        this.profile = in.readBoolean();
        this.queryId = in.readOptionalString();
    }

    @Override
    public void writeTo(StreamOutput out) throws IOException {
        super.writeTo(out);
        out.writeOptionalString(pplQuery);
        out.writeOptionalString(format);
        out.writeOptionalString(explainMode);
        out.writeOptionalString(jsonContentString);
        out.writeOptionalString(path);
        out.writeBoolean(sanitize);
        out.writeEnum(style);
        out.writeBoolean(profile);
        out.writeOptionalString(queryId);
    }

    public String getPplQuery() {
        return pplQuery;
    }

    public String getJsonContentString() {
        return jsonContentString;
    }

    public String getPath() {
        return path;
    }

    public String getFormat() {
        return format;
    }

    public String getExplainMode() {
        return explainMode;
    }

    public boolean getSanitize() {
        return sanitize;
    }

    public Style getStyle() {
        return style;
    }

    public boolean getProfile() {
        return profile;
    }

    public String getQueryId() {
        return queryId;
    }

    @Override
    public ActionRequestValidationException validate() {
        return null;
    }
}

File: src/main/kotlin/org/opensearch/commons/sql/action/TransportPPLQueryRequest.kt
/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

package org.opensearch.commons.sql.action

import org.opensearch.action.ActionRequest
import org.opensearch.action.ActionRequestValidationException
import org.opensearch.action.ValidateActions
import org.opensearch.common.io.stream.StreamInput
import org.opensearch.common.io.stream.StreamOutput
import org.opensearch.common.io.stream.Writeable
import org.opensearch.common.xcontent.ToXContent
import org.opensearch.common.xcontent.ToXContentObject
import org.opensearch.common.xcontent.XContentBuilder
import org.opensearch.commons.sql.model.Style
import java.io.IOException

/**
 * Action Request to send ppl query.
 */
class TransportPPLQueryRequest : ActionRequest, ToXContentObject {
    val pplQuery: String
    val path: String
    var format: String
    var sanitize: Boolean?
    var style: Style?
    val threadContext: String?

    companion object {
        /**
         * reader to create instance of class from writable.
         */
        val reader = Writeable.Reader { TransportPPLQueryRequest(it) }
    }

    /**
     * constructor for creating the class
     * @param query the ppl query string
     * @param threadContext the user info thread context
     */
    constructor(
        query: String,
        path: String,
        format: String = "",
        sanitize: Boolean? = true,
        style: Style? = Style.COMPACT,
        threadContext: String?
    ) {
        this.pplQuery = query
        this.path = path
        this.format = format
        this.sanitize = sanitize
        this.style = style
        this.threadContext = threadContext
    }

    /**
     * {@inheritDoc}
     */
    @Throws(IOException::class)
    constructor(input: StreamInput) : super(input) {
        pplQuery = input.readString()
        path = input.readString()
        format = input.readString()
        sanitize = input.readOptionalBoolean()
        style = input.readEnum(Style::class.java)
        threadContext = input.readOptionalString()
    }

    @Throws(IOException::class)
    override fun writeTo(output: StreamOutput) {
        super.writeTo(output)
        output.writeString(pplQuery)
        output.writeString(path)
        output.writeOptionalString(format)
        output.writeOptionalBoolean(sanitize)
        output.writeEnum(style)
        output.writeOptionalString(threadContext)
    }

    /**
     * {@inheritDoc}
     */
    override fun toXContent(builder: XContentBuilder?, params: ToXContent.Params?): XContentBuilder? {
        return null
    }

    /**
     * {@inheritDoc}
     */
    override fun validate(): ActionRequestValidationException? {
        var validationException: ActionRequestValidationException? = null
        if (pplQuery.isEmpty()) {
            validationException = ValidateActions.addValidationError("query is empty", validationException)
        }
        return validationException
    }
}
